Change how `numpy` behaves as an optional import

jagerber48 opened the pull request:
- [x] Closes #267
- [x] Executed `pre-commit run --all-files` with no errors
- [x] The change is fully covered by automated unit tests
- [x] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

wshanks commented on uncertainties/core.py:
```diff
@@ -65,139 +66,151 @@
-    # Entering variables as a block of correlated values.  Only available
-    # if NumPy is installed.
+    msg = "Unable to import numpy. Some functionality will be unavailable"
+    warnings.warn(msg)
```
If I were using Uncertainties with Numpy, this warning every time I did `import uncertainties` would annoy me. It is better to defer the warning until the first attempt to use an unsupported feature (at which point it will need to be an exception instead of a warning).

wshanks commented on uncertainties/core.py:
```diff
@@ -65,139 +66,151 @@
-        Return numbers with uncertainties (AffineScalarFunc objects)
-        that correctly reproduce the given covariance matrix, and have
-        the given (float) values as their nominal value.
+def correlated_values(nom_values, covariance_mat, tags=None):
```
I think there are two ways we can change things here:

1. Put some check at the top of the function to make sure that `numpy` was imported correctly and raise an exception if not.
2. Conditionally define the functions like was done before but in the else case instead of not defining the functions define functions with the same name and signature but with bodies that just raise an exception about numpy missing.

The first option seems simpler. It just has that extra check at the start of every call but that should be okay.

For the exception I think it would be a custom exception, or maybe `ModuleNotFoundError`, though it feels a little weird to raise an `ImportError` during a function call.

Minor point -- maybe the `try` should be excepting `ModuleNotFoundError` instead of `ImportError`, just in case `numpy` is installed but there is some other problem with importing it? Right now, Uncertainties hides that other import error from the user just like if numpy were not installed.

newville commented:
@jagerber48 @wshanks I agree about avoiding the warning for the failed import.   I think raising an Exception in the function would be best, probably a plain `Exception` or `NotImplementedError` (ie, it is not implemented without numpy), but not an `ImportError`.

I might say that this is a good argument for making numpy a hard requirement (discussed also in #227, where the general consensus was to not add it as a requirement).

wshanks commented:
I will say on my part that I like the idea of Numpy being an optional dependency (particularly since there are no dependencies), but in practice I don't have any use cases for Uncertainties that don't also involve Numpy being installed. It might be interesting to explore the GitHub search a little bit for requirements.txt, setup.py, etc. and see if it seems like using Uncertainties without Numpy has a user base. (@lebigot might also know of some cases).

newville commented:
@wshanks I agree that going from 0 dependencies to 1 is a big (~infinite!) leap.   

I should say that I do maintain one completely unrelated library that is also "will use Numpy if available, but it is not required".     Like `uncertainties`, that code is 10+ years old, and Numpy seemed much less ubiquitous at that time (and wheels were not a thing, I think;)).  

Anyway, I'm not quite advocating for "add Numpy dependency", just pointing out that we probably really could do that without causing much pain.

jagerber48 commented on uncertainties/core.py:
```diff
@@ -65,139 +66,151 @@
+        "correlated_values",
+        "correlated_values_norm",
+        "correlation_matrix",
+    ]
```
Yeah, in a separate PR we should get rid of that `from .core import *` and just explicitly define the public API there. It will help with various automated code inspections.

jagerber48 commented:
Ok,

- Removed the warning at import time
- Added a `NotImplementedError` at function call time. Let me know if there are recommendations for improved wording for the error messages.
- Add the functions to `__all__` unconditionally.

Not sure what the deal is with codecov. I'll look into it.

________

One use case for `uncertainties` without `numpy` is if you are using `uncertainties` purely for formatting which I think some people do. I have done it on occasion. However, because of the work I do I am **always** using `numpy`. That is to say, there are times when I use `uncertainties` but don't require the `numpy` parts of `uncertainties`, but unrelated I do have `numpy` installed. I would be in favor of making `numpy` a required dependency because I doubt it would cause anyone trouble. But I would be interested in seeing the study of `requirements.txt`, `setup.py`, `pyproject.toml` etc. looking for dependency pairings. Regardless, not trying to stir the pot on that issue anyways. Happy to leave as is and proceed with this PR.

wshanks reviewed (approved):
Looks good to me!

jagerber48 commented on tests/test_uncertainties.py:
```diff
@@ -1313,3 +1321,39 @@ def test_correlated_values_correlation_mat():
         )
+
+
+@patch("uncertainties.core.numpy", None)
```
Yeah, I wasn't sure a good way to do this. The issue is that the added lines were counting as drops in test coverage. Do you know if the no-numpy CI job contributes to codecov.io test coverage?

jagerber48 commented:
After the last failed test I think I will just try to leverage the no numpy environment. That is the most accurate test scenario anyways for what I'm trying to test. Hopefully the results get bundled into code coverage appropriately.

wshanks commented:
> After the last failed test I think I will just try to leverage the no numpy environment. That is the most accurate test scenario anyways for what I'm trying to test. Hopefully the results get bundled into code coverage appropriately.

Okay. I tried to set up the no numpy test so that it would be added to the coverage, but I haven't looked carefully at the report to check that. It should be possible to see, looking at the branches for when numpy is not installed.

wshanks commented:
It looks like the `pass` statement [here](https://app.codecov.io/gh/lmfit/uncertainties/commit/9a4c27528e597c6a9320094e40063e173b0d908d/blob/uncertainties/core.py#L69) for example is covered.

jagerber48 commented:
I'll try to get to it today. If not it may have to wait until next week.

jagerber48 commented:
Ok, so I generally like testing this way. Skipping the "no numpy behavior" tests when numpy is present. However, it's not working with codecov. On another package I hit issues when I was uploading multiple reports to codecov. Like how we're uploading to codecov for all versions and operating systems and also for the no numpy test. One thing I did was implement flags. This way the different runs get submitted with different flags. The different flags have different coverage, but the total coverage is the union over all of the different flags. This resolved some problems like this for me. There's also something called "components". I can't tell exactly how flags and components are used on codecov, but it's stuff to look into I guess

wshanks commented:
Huh, that is odd. I don't have much experience with Codecov. There are probably tweaks we can do to make it better. For some reason the run for the current commit on master does show the no-numpy lines as covered in Codecov. When you go to the Codecov page for a commit, there is a "Coverage reports history" box to the side with 16 xml files. They all have the same label (one of the tweaks I was thinking of was labeling these by environment). Clicking through and scanning each one I found that the no-numpy case is one of the 16 both for 9a4c275 (master) and bce701B (latest commit here) and does show the `except` side of the `try: import numpy` covered even though the Codecov page for bce701B shows that part as not covered. So if there is a way to tweak Codecov to combine them more reliably that would be good.

An alternative would be to pass the reports through a separate workflow and use coverage to combine them into one XML file that gets uploaded to Codecov. Hopefully we don't have to do that.

I don't think any of this is critical but it would be nice to take the time to get it right.

newville commented:
@wshanks @jagerber48 I waited a bit to respond, being somewhat disillusioned with the whole concept of coverage and the game-ification that Github seems to encourage.  Can I encourage focusing on "does the code do what we want" rather than "passing" tests and improving some made up "score"?

wshanks commented:
@newville I think the code does what we want and so I approved the PR. I don't understand the Codecov behavior. The coverage it reports overall does not match the union of the reports from the individual test runs though the Codecov documentation says that it should. I don't think getting Codecov right is that important, but also Uncertainties has handled Numpy the same way for many years, so I didn't think there was any rush to merge this PR. For this reason, I wrote "I don't think any of this is critical but it would be nice to take the time to get it right."

I think the coverage reports are a moderately useful tool for checking that the tests did what they were intended to do and the tests are similarly a moderately useful tool for checking that the code did what it was intended to do. I understand you want visibility into the project as a maintainer and so want to follow all activity on the repo, and so any aspect that you don't find valuable becomes tedious for others to have long discussions about. Perhaps it is best to merge this PR in its current form, and then if @jagerber48 wants to sort out the Codecov configuration he could open a separate Codecov-only PR which you could unsubscribe from. I don't mind reviewing though I am not motivated enough myself to sort it out. As this is a volunteer effort, I don't think we should dictate what the other maintainers should find worth spending time on, though we should be respectful of all maintainers' time and try to limit noise in the repo activity feed.

jagerber48 commented:
I've spent time with codecov, it can be made to work in nice ways, but it does take effort to maintain and there's a disappointingly large learning curve. I like having CI that measures and reports coverage. I've found that the target of high code coverage has served to improve my codebases. That said, I understand the frustration when "checking the box" unnecessarily throws a wrench in the gears towards making progress like it is here.

I think a solution I would be happy with would be to continue to measure and report code coverage, but to remove it as a check on merging PRs. So humans can look at the code coverage and react accordingly, but a small drop (or any drop) in code coverage won't show up as a failed test. Idea that crossed my mind but I wouldn't push for unless others are interested: We could add a manual checkbox to the PR template asking the PR author if the code coverage is maintained or increased by the PR. If it is decreased they can comment as to why that is ok (probably something like, the loss in coverage is minor/insignificant) or a discussion can be had about it.

jagerber48 commented:
Ok, this PR is ready for review after the codecov stuff was sorted out by #269. 

There are still some annotations about lines not being covered appearing in the github diff, but these are erroneous. They are in fact covered, and all the other codecov reports indicate so. I can't quite get totally get them to clear but they can be safely ignored.

Either those sorts of issues will go away after a couple PRs go through, or we need to add `codecov.notify.after_n_builds` to the `.codecov.yaml` to make sure `codecov.io` isn't sending a premature erroneous notification to github before all reports are uploaded.

newville commented:
@jagerber48 @wshanks Great!  Github wants this to be "squashed and merged", so I'll do that.
